### DISCUSSION TITLE
The Bug that Turned into a Beautiful Feature: Structured Legend Parent/Child Symbology Handling

### DISCUSSION THREAD

**Discussion by mweech** [fgpv-vpgf/fgpv-vpgf#4115] [Team Posts]
Cross post of initial comment by @AleksueiR for discussion: https://github.com/fgpv-vpgf/fgpv-vpgf/issues/3152#issuecomment-445801582

At this point, I think everyone agrees these two use cases are worthwhile and should be addressed. Regardless of our initial intentions, the user-expected behaviour needs to be preserved; furthermore, it matches the default behaviour of the ArcGIS Online map viewer which many RAMP users will be familiar with.

I can think of three options:
#### 1. Legacy mode
A new config option is added to toggle between `legacy` (children's states are preserved upon toggling their parent) and `new` (children's states are reset to their parent's state) modes of parent/child interaction. A config author would need to select which state the map will operate in.

Pros:
- separates our botched implementation from initial intentions
- provides a mode aligning with ArcGIS Viewer map
- provides a mode implementing our initial intentions

Cons:
- need to implement both approaches
- a map author needs to know about this option and decide which mode to use
- a user would not know which mode is active (this one is really bad - maybe a mode indicator should be added anyway, or the user _can_ be given control over the mode) and would not have any control over it
- the same user might encounter maps with both modes with would be confusing

#### 2. Hybrid
We implement the legacy mode (children's states are preserved) and add additional two options to all groups' context menus letting users manually reset the group and all its children to visible or hidden. 

![](https://i.imgur.com/eRir3lu.png)

This will be similar to the visibility menu we have in the top right corner of the panel which can show/hide all the layers at once.

![](https://i.imgur.com/T8zbCp2.png)

Pros:
- need to implement only one approach
- users are likely to be familiar with it from 2.4, so there is no learning curve
- users have new menu options to reset groups' visibility similar to our original intentions
- aligns with behaviour of the ArcGIS Map viewer and exclusive visibility group (https://github.com/fgpv-vpgf/fgpv-vpgf/issues/2224#issuecomment-316987070)

Cons:
- these menu options are not dead obvious since they are hidden in the menu, but they are discoverable
- most work (arguably)

#### 3. Lazy
Implement just the legacy mode and pretend nothing happened.

Pros:
- least amount of work (arguably)
- users are likely to be familiar with it from 2.4, so there is no learning curve
- aligns with behaviour of the ArcGIS Map viewer and exclusive visibility group (https://github.com/fgpv-vpgf/fgpv-vpgf/issues/2224#issuecomment-316987070)

Cons:
- resetting visibility for a large group needs to be done by hand which can be tedious for the user

### Thoughts
Option 1 is the worst since we start adding invisible app modes which users can't control (or worse, _can_ control) - different application modes just needlessly tax users' memory and attention.

Option 2 is the most favourable, in my opinion, since we preserve the behaviour users are familiar with and provide an option to reset group visibility through the menu. 

Option 3 is sub-optimal and it's technically a half of Option 2 and can be implemented if there is no time to fully implement Option 2.

### Note
The 2.4 implementation preserves the states only of the symbology items (as far as I know). In options 1 and 2, this needs to be expanded to also preserve the state of actual child element (`LegendNode`s) as well. Otherwise, the behaviour becomes fragmented.

**Comment by spencerwahl** [comment]
My vote is for option 2, I think it provides the best user experience. I also think the buttons in the menu are discover-able enough; adding the eyeball toggle icons to the group item (not in the menu) would probably cause more confusion.

**Comment by dane-thomas** [comment]
I agree with what Spencer said to use option 2 but just adding the toggle to the existing menu.

In terms of amount of work, the code added to fix the bug initially is in one spot so removing should be straight forward. As well, some of that code can be adapted for show/hide all buttons.

**Comment by james-rae** [comment]
Looking like we have consensus with Option 2. Well done!  Will progress to making Implementation Details and Acceptance Criteria
